toolchain/check/cpp_overload_resolution.cpp

```diff
@@ -0,0 +1,111 @@
+        // TODO: Add location accordingly.
+        clang::SourceLocation(), arg_cpp_type->getNonReferenceType(),
+        clang::ExprValueKind::VK_LValue);
+    arg_exprs.emplace_back(arg_expr);
+  }
+
+  auto fn_type_inst =
+      context.types().GetAsInst(context.insts().Get(callee_id).type_id());
+  auto fn_type = fn_type_inst.TryAs<SemIR::OverloadedCppFunctionType>();
+  if (!fn_type) {
+    return std::nullopt;
+  }
+  const SemIR::OverloadedCppFunction& overloaded_fn =
+      context.overloaded_cpp_functions().Get(fn_type->overloaded_function_id);
+
+  // Add candidate functions from the name lookup.
+  clang::OverloadCandidateSet candidate_set(
+      // TODO: Add location accordingly.
+      clang::SourceLocation(),
+      clang::OverloadCandidateSet::CandidateSetKind::CSK_Normal);
+
+  clang::ASTUnit* ast = context.sem_ir().clang_ast_unit();
+  CARBON_CHECK(ast);
+  clang::Sema& sema = ast->getSema();
+
+  for (clang::NamedDecl* candidate : overloaded_fn.candidate_functions) {
+    if (auto* fn_decl = dyn_cast<clang::FunctionDecl>(candidate)) {
+      sema.AddOverloadCandidate(
+          fn_decl, clang::DeclAccessPair::make(fn_decl, candidate->getAccess()),
+          arg_exprs, candidate_set);
+    }
+  }
+
+  // Find best viable function among the candidates.
+  // Note: In C++, a single non-templated function is also treated as an
+  // overloaded set and goes through the overload resolution to ensure that the
+  // function is viable for the call. Keeping the same behavior here for
+  // consistency.
+  clang::OverloadCandidateSet::iterator best_viable_fn;
+  clang::OverloadingResult overloading_result =
+      // TODO: Add location accordingly.
+      candidate_set.BestViableFunction(sema, clang::SourceLocation(),
+                                       best_viable_fn);
+
+  switch (overloading_result) {
+    case clang::OverloadingResult::OR_Success: {
+      SemIR::InstId result =
+          ImportDeclAndDependencies(context, loc_id, best_viable_fn->Function);
+      return result;
+    }
+    case clang::OverloadingResult::OR_No_Viable_Function: {
+      CARBON_DIAGNOSTIC(
+          OverloadingNoViableFunctionFound, Error,
+          "no viable function found during overloading resolution");
+      context.emitter().Emit(loc_id, OverloadingNoViableFunctionFound);
+      return std::nullopt;
+    }
+    case clang::OverloadingResult::OR_Ambiguous: {
+      CARBON_DIAGNOSTIC(
+          OverloadingAmbiguousCandidatesFound, Error,
+          "ambiguous candidates found during overloading resolution");
+      context.emitter().Emit(loc_id, OverloadingAmbiguousCandidatesFound);
+      return std::nullopt;
+    }
+    case clang::OverloadingResult::OR_Deleted: {
+      CARBON_DIAGNOSTIC(
+          OverloadingDeletedFunctionFound, Error,
+          "overloading resolution succeeded, but refers to a deleted function");
+      context.emitter().Emit(loc_id, OverloadingDeletedFunctionFound);
```

Comment:
```suggestion
      CARBON_DIAGNOSTIC(
          OverloadingNoViableFunctionFound, Error,
          "no matching function for call to `{0}`", SemIR::NameId);
      context.emitter().Emit(loc_id, OverloadingNoViableFunctionFound, overloaded_fn.name_id);
      return std::nullopt;
    }
    case clang::OverloadingResult::OR_Ambiguous: {
      CARBON_DIAGNOSTIC(
          OverloadingAmbiguousCandidatesFound, Error,
          "call to `{0}` is ambiguous", SemIR::NameId);
      context.emitter().Emit(loc_id, OverloadingAmbiguousCandidatesFound, overloaded_fn.name_id);
      return std::nullopt;
    }
    case clang::OverloadingResult::OR_Deleted: {
      CARBON_DIAGNOSTIC(
          OverloadingDeletedFunctionFound, Error,
          "call to deleted function `{0}`", SemIR::NameId);
      context.emitter().Emit(loc_id, OverloadingDeletedFunctionFound, overloaded_fn.name_id);
```
I think it'd make sense to use the same wording for these diagnostics that Clang does.

Reply:
Done.